HPS/Heartland/Model/Payment.php

```diff
@@ -368,129 +275,629 @@ public function refund(\Magento\Payment\Model\InfoInterface $payment, $amount)
+                $validCardHolder = $this->getHpsCardHolder($order->getBillingAddress());
+
+                $this->log($paymentAction, 'HPS\Heartland\Model\Payment $paymentAction: ');
+                if ($paymentAction === \HpsTransactionType::AUTHORIZE || $paymentAction === \HpsTransactionType::CHARGE) {
+                    $this->log($suToken, 'HPS\Heartland\Model\Payment getToken Method Called: ');
+                    // \HPS\Heartland\Model\Payment::$_token_value
+
+                    $suToken
+                        = $this->getToken(new \HpsTokenData,
+                                          $order->getCustomerId()); //$this->getSuToken();// this just gets the passed
+                    // token value
+
+                    $this->log($suToken, 'HPS\Heartland\Model\Payment after getToken Method Called: ');
+                }
+            }
+
+            /*
+             * execute the portic messages related to the specified action
+             */
+            switch ($paymentAction) {
+                /*
+                 * \HpsTransactionType::AUTHORIZE places a hold on the card and requests an approval from the Card
+                 * Issuer
+                 * This transaction will not be settled and the hold will fall off of the card holders account
+                 * usually after 7 days but this time frame is up to the issuer and may vary.
+                 * If the transaction is later captured it is then transfered from the card holders account to the
+                 * merchant.
+                 * Approval codes are typically only good for 30 days or so
+                 *
+                 * the typical use case for this transaction is if a product is ordered and not immediately shipped
+                 */
+                case (\HpsTransactionType::AUTHORIZE): // Portico CreditAuth \HpsTransactionType::AUTHORIZE
+
+                    $this->log($suToken, 'HPS\Heartland\Model\Payment authorize Method Called: ');
+                    /** @var \HpsAuthorization $response Properties found in the HpsAuthorization */
+                    ;
+                    $response = $chargeService->authorize(\HpsInputValidation::checkAmount($requestedAmount),
+                                                          $currency,
+                                                          $suToken,
+                                                          $validCardHolder,
+                        ($this->validateSuToken() ? true : false),
+                                                          null,
+                                                          $storeName);
+                    if (isset($response->tokenData) && $response->tokenData->tokenValue){
+                        $payment->setCcNumberEnc($response->tokenData->tokenValue);
+                    }
+                    if(!$this->validateSuToken()){
+                        $payment->setCcNumberEnc($suToken->tokenValue);
+                    }
```

Comment:
1) logic for multi use token is updated
2) Authorize calls $payment->setCcNumberEnc stores the token value to capture
3) Charge calls $payment->setParentTransactionId to update the transaction is completed